File: Propositions/Book1/Prop01_equilateral_triangle.pl
#!/usr/bin/perl
use strict;
use warnings;

use Geometry::Canvas::PropositionCanvas;
use FindBin;
use lib "$FindBin::Bin/..";
use Proposition;

my $pn = PropositionCanvas->new;
Proposition::init($pn);
Proposition::set_animation(0);

# ============================================================================
# Definitions
# ============================================================================
my $title =
  "To construct an equilateral triangle on a given finite straight line.";

my $t1 = $pn->text_box( 800, 150, -width => 500 );
my $t2 = $pn->text_box( 500, 430 );
$pn->title(1,$title);

my $steps;
push @$steps, Proposition::title_page($pn);
push @$steps, Proposition::toc($pn,1);
push @$steps, Proposition::reset();
explanation( $pn );
push @$steps,sub{Proposition::last_page};
$pn->define_steps($steps);
$pn->copyright();

$pn->go;



# ============================================================================
# Explanation
# ============================================================================
sub explanation {

    my ( %l, %p, %c );

    my @A = ( 200, 500 );
    my @B = ( 450, 500 );

    # -------------------------------------------------------------------------
    # Construction
    # -------------------------------------------------------------------------
    push @$steps, sub {
        $t1->title("Construction:");
        $t1->explain("Start with line segment AB");
        Point->new($pn,@A)->label( 'A', 'left' );
        $l{AB} = Line->new($pn, @A, @B );
        Point->new($pn,@B)->label( 'B', 'right' );
    };

    push @$steps, sub {
        $t1->explain("Create a circle with center A and radius AB");
        $c{A} = Circle->new($pn, @A, @B );
    };

    push @$steps, sub {
        $t1->explain("Create a circle with center B and radius AB");
        $c{B} = Circle->new($pn, @B, @A );
    };

    push @$steps, sub {
        $t1->explain("Label the intersection point C");
        my @ps = $c{A}->intersect( $c{B} );
        $p{C} = Point->new($pn, @ps[ 0, 1 ] )->label( "C", 'top' );
    };

    push @$steps, sub {
        $t1->explain("Create line AC and CB");
        $l{AC} = Line->new($pn, @A,            $p{C}->coords );
        $l{BC} = Line->new($pn, $p{C}->coords, @B );
    };

    push @$steps, sub {
        $t1->explain("Triangle ABC is an equilateral triangle");
        $c{A}->grey;
        $c{B}->grey;
    };

    # -------------------------------------------------------------------------
    # Proof
    # -------------------------------------------------------------------------
    push @$steps, sub {
        $t1->down;
        $t1->title("Proof:");
        $l{AC}->grey;
        $c{A}->grey;
        $c{B}->normal;

        $t1->explain(
            "AB and CB are radii of the same circle - hence they are equal");
        $l{AB}->label( "r", "bottom" );
        $l{BC}->label( "r", "right" );
        $t2->math("AB = CB = r");
    };

    push @$steps, sub {
        $l{BC}->grey;
        $c{B}->grey;
        $l{AB}->normal;
        $l{AC}->normal;
        $c{A}->normal;

        $t1->explain(
            "AB and AC are radii of the same circle - hence they are equal");
        $l{AC}->label( "r", "left" );
        $t2->math("AB = AC = r");
    };

    push @$steps, sub {
        $l{AC}->normal;
        $l{AB}->normal;
        $l{BC}->normal;
        $c{B}->grey;
        $c{A}->grey;

        $t1->explain(
                "If AB equals AC and AB equals CB, "
              . "then AC equals CB"
        );
        $l{BC}->label( "r", "right" );
        $t2->math("AB = CB = CA = r");
    };

    push @$steps, sub {
        $t2->down;
        $t2->math("\\{therefore} Equilateral Triangle!");

    };
    return $steps;

}

File: Propositions_python_code/Book1/Prop01_equilateral_triangle.py
import sys
import os
sys.path.append(os.getcwd())

from euclidlib.Scenes.book01 import Book1Scene
from euclidlib.Objects import *


class Book1Prop1(Book1Scene):
    title = "To construct an equilateral triangle on a given finite straight line."
    steps = []

    def go(self):
        t1 = TextBox(mn_coord(800, 150), line_width=mn_scale(500))
        t2 = TextBox(mn_coord(500, 430))
        A = mn_coord(200, 500)
        B = mn_coord(450, 500)

        l: dict[str | int, ELine] = {}
        p: dict[str | int, EPoint] = {}
        c: dict[str | int, ECircle] = {}

        # -------------------------------------------------------------------------------------------------------------
        t1.title("Construction:")
        t1.explain("Start with line segment AB")
        p['A'] = EPoint(A, scene=self, label=('A', LEFT))
        p['B'] = EPoint(B, scene=self, label=('B', RIGHT))
        l['AB'] = ELine(p['A'], p['B'], scene=self)

        # -------------------------------------------------------------------------------------------------------------
        self.next_page()
        t1.explain("Create a circle with center A and radius AB")
        c['A'] = ECircle(p['A'], p['B'], scene=self)

        # -------------------------------------------------------------------------------------------------------------
        self.next_page()
        t1.explain("Create a circle with center B and radius AB")
        c['B'] = ECircle(p['B'], p['A'], scene=self)

        # -------------------------------------------------------------------------------------------------------------
        self.next_page()
        t1.explain("Label the intersection point C")
        pts = c['A'].intersect(c['B'])
        p['C'] = EPoint(pts[0], scene=self, label=('C', UP))

        # -------------------------------------------------------------------------------------------------------------
        self.next_page()
        t1.explain("Create line AC and CB")
        with self.simultaneous():
            l['AC'] = ELine(p['A'], p['C'], scene=self)
            l['BC'] = ELine(p['B'], p['C'], scene=self)

        # -------------------------------------------------------------------------------------------------------------
        self.next_page()
        t1.explain("Triangle ABC is an equilateral triangle")
        with self.simultaneous():
            c['A'].e_fade()
            c['B'].e_fade()

        # -------------------------------------------------------------------------------------------------------------
        # Proof
        # -------------------------------------------------------------------------------------------------------------
        self.next_page()
        t1.down()
        t1.title("Proof:")
        with self.simultaneous():
            l['AC'].e_fade()
            c['A'].e_fade()
            c['B'].e_normal()

        t1.explain("AB and CB are radii of the same circle - hence they are equal")
        l['AB'].add_label("r", DOWN)
        l['BC'].add_label("r", RIGHT)
        t2.math("AB = CB = r")

        # -------------------------------------------------------------------------------------------------------------
        self.next_page()
        with self.simultaneous():
            l['BC'].e_fade()
            c['B'].e_fade()
            l['AB'].e_normal()
            l['AC'].e_normal()
            c['A'].e_normal()

        t1.explain("AB and AC are radii of the same circle - hence they are equal")
        l['AC'].add_label("r", LEFT)
        t2.math("AB = AC = r")

        # -------------------------------------------------------------------------------------------------------------
        self.next_page()
        with self.simultaneous():
            c['B'].e_fade()
            c['A'].e_fade()
            l['AB'].e_normal()
            l['AC'].e_normal()
            l['BC'].e_normal()

        t1.explain(
            "If AB equals AC and AB equals CB, "
            "then AC equals CB"
        )
        l['BC'].add_label("r", RIGHT)
        t2.math("AB = CB = CA = r")

        # -------------------------------------------------------------------------------------------------------------
        self.next_page()
        t2.down()
        t2.math(r"\therefore\quad Equilateral\quad Triangle!")

